### PR TITLE
Usps address verification

### LINKED ISSUE #5099: usps address verification
### Describe the problem

currently isn't a way to hit the usps api to check on an address

### Describe the solution you'd like

a minimal viable product that verifies demographics contact address with usps web api tools

### PR DESCRIPTION
**Pull request by stephenwaite**
Fixes #5099 

#### Short description of what this resolves:
verifies demographics contact address with usps web api

#### Changes proposed in this pull request:
added some code to implement

Screenshot from 2022-04-06 14-11-00

### CODE REVIEW

File: interface/patient_file/summary/demographics_full.php [hunk 1]
```diff
@@ -312,6 +312,20 @@ function checkNum () {
  }
 }
 
+function address_verify() {
+    var f = document.demographics_form;
+
```
Comment (bradymiller) [inline interface/patient_file/summary/demographics_full.php:317]: top.restoreSession();

File: interface/patient_file/summary/demographics_full.php [hunk 2]
```diff
@@ -312,6 +312,20 @@ function checkNum () {
...
+function address_verify() {
+    var f = document.demographics_form;
+    
+    dlgopen('../../practice/address_verify.php?address1=' + encodeURIComponent(f.form_street.value) +
+    '&address2=' + encodeURIComponent(f.form_street_line_2.value) +
+    '&city=' + encodeURIComponent(f.form_city.value) +
+    '&state=' + encodeURIComponent(f.form_state.value) +
+    '&zip5=' + encodeURIComponent(f.form_postal_code.value.substring(0,5)) +
+    '&zip4=' + encodeURIComponent(f.form_postal_code.value.substring(5,9))
+    , '_blank', 400, 150, '','Address Verify');
```
Comment (bradymiller) [inline interface/patient_file/summary/demographics_full.php:324]: translate Address Verify

File: interface/patient_file/summary/demographics_full.php [hunk 3]
```diff
@@ -1146,6 +1160,11 @@ class='text'>(<?php echo xlt('Browse'); ?>)</a>
         $(".sel2").select2({
             <?php require($GLOBALS['srcdir'] . '/js/xl/select2.js.php'); ?>
         });
+
+        <?php if ($GLOBALS['usps_webtools_enable']) { ?>
+            $("#value_id_text_postal_code").append(
+                "<input type='button' class='btn btn-sm btn-secondary mb-1' onclick='address_verify()' value ='Verify Address' />");
```
Comment (bradymiller) [inline interface/patient_file/summary/demographics_full.php:1166]: translate Address Verify

File: interface/practice/address_verify.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            $output .= "<div class='text-danger'>";
+        } else {
+            $output .= "<div class='text-success'>";
+        }
+        $output .= text($key) . ": " . text($value) . "</div>";
+    }
+    //$output = var_dump($response_array);
+} else {
+    $output .= "<div class='text-danger'>";
+    $output .= 'Error: ' . $verify->getErrorMessage()  . "</div>";
```
Comment (bradymiller) [inline interface/practice/address_verify.php:85]: text()

File: src/USPS/USPSBase.php [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+ * @author Vincent Gabriel
+ */
+class USPSBase
+{
+    const LIVE_API_URL = 'https://secure.shippingapis.com/ShippingAPI.dll';
+
+  /**
+   * @var string - the usps username provided by the usps website
+   */
+    protected $username = '633CAREM3197';
```
Comment (bradymiller) [inline src/USPS/USPSBase.php:35]: is this supposed to be hard coded?

**Comment by bradymiller** [comment]
review done (didn't test)
neat stuff!

2DV

**Comment by stephenwaite** [comment]
thanks for the review @bradymiller ! 

it's testing well and bringing in, will write a short instruction for signing up with the usps web api in the wiki too

**Pull request merged**